compiler/rustc_next_trait_solver/src/solve/eval_ctxt/mod.rs

```diff
@@ -848,6 +848,16 @@ where
         self.delegate.enter_forall(value, f)
     }
 
+    /// `enter_forall`, but takes `&mut self` and passes it back through the
+    /// callback since it can't be aliased during the call.
+    pub(super) fn enter_forall_mut<T: TypeFoldable<I> + Copy, U>(
+        &mut self,
+        value: ty::Binder<I, T>,
+        f: impl FnOnce(&mut Self, T) -> U,
+    ) -> U {
+        self.delegate.enter_forall(value, |value| f(self, value))
```

Comment:
given that `enter_forall` is only used once, maybe it's easier to always use `enter_forall_mut` by modifying the signature of `neter_forall`.

While you're at it, change `fn compute_goal` to use `self.enter_forall` from `self.delegate.enter_forall` :grin:

Reply:
I remember there being a specific problem where we had an infectious &mut self that ended up changing quite a lot if we did this

Reply:
nvm, just needed to move an arg lol